extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java

```diff
@@ -1,189 +1,266 @@
+          AsyncKuduScanner scanner =
+              scannerBuilder.addPredicate(lowerPred).addPredicate(upperPred).addPredicate(
+                  partitionPred).addPredicate(adapterIdPred).build();
+          scanners.add(scanner);
         }
-      } else {
-        KuduScannerBuilder scannerBuilder = operations.getScannerBuilder(table);
-        KuduScanner scanner = scannerBuilder.addPredicate(adapterIdPred).build();
-        KuduUtils.executeQuery(scanner, results);
       }
+    } else {
+      AsyncKuduScannerBuilder scannerBuilder = operations.getAsyncScannerBuilder(table);
+      AsyncKuduScanner scanner = scannerBuilder.addPredicate(adapterIdPred).build();
+      scanners.add(scanner);
     }
 
-    final Iterator<GeoWaveRow> tmpIterator;
-    Iterator<RowResult> concatIterator = Iterators.concat(results.iterator());
-    if (dataIds == null) {
-      Function<RowResult, GeoWaveRow> adapter;
-      if (isDataIndex) {
-        adapter = r -> KuduDataIndexRow.deserializeDataIndexRow(r, visibilityEnabled);
-      } else {
-        adapter = KuduRow::new;
+    return executeQueryAsync(scanners.toArray(new AsyncKuduScanner[] {}));
+  }
+
+  public CloseableIterator<T> executeQueryAsync(final AsyncKuduScanner... scanners) {
+    final BlockingQueue<Object> results = new LinkedBlockingQueue<>(MAX_BOUNDED_READS_ENQUEUED);
+    final AtomicBoolean isCanceled = new AtomicBoolean();
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        final AtomicInteger queryCount = new AtomicInteger(1);
+        for (final AsyncKuduScanner scanner : scanners) {
+          try {
+            readSemaphore.acquire();
+            executeScanner(
+                scanner,
+                readSemaphore,
+                results,
+                queryCount,
+                isCanceled,
+                visibilityEnabled,
+                filter,
+                rowTransformer,
+                rowMerging,
+                adapterIds);
+          } catch (final InterruptedException e) {
+            LOGGER.warn("Exception while executing query", e);
+            readSemaphore.release();
+          }
+        }
+        // then decrement
+        if (queryCount.decrementAndGet() <= 0) {
+          // and if there are no queries, there may not have
+          // been any
+          // statements submitted
+          try {
+            results.put(RowConsumer.POISON);
+          } catch (final InterruptedException e) {
+            LOGGER.error(
+                "Interrupted while finishing blocking queue, this may result in deadlock!");
+          }
+        }
       }
-      if (visibilityEnabled) {
-        tmpIterator = Streams.stream(concatIterator).map(adapter).filter(filter).iterator();
-      } else {
-        tmpIterator = Iterators.transform(concatIterator, adapter);
+    }, "Kudu Query Executor").start();
+    return new CloseableIteratorWrapper<T>(new Closeable() {
```

Comment:
Could you also replace this anonymous `Closeable` with a lambda?

Reply:
It doesn't allow me to use void as the return type for close() if I do that

Reply:
You should be able to replace this with: `return new CloseableIteratorWrapper<T>(() -> isCanceled.set(true), new RowConsumer(results));`